ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -139,134 +132,117 @@ private[kvutils] class ProcessTransactionSubmission(
+                  && value.getContractState.hasContractKey
+                  && contractIsActiveAndVisibleToSubmitter(digest, value.getContractState) =>
+              value.getContractState.getContractKey -> Conversions.stateKeyToContractId(key)
+          }
 
-    engine
-      .validate(
-        transactionEntry.abs,
-        transactionEntry.ledgerEffectiveTime,
-        participantId,
-        transactionEntry.submissionTime,
-        transactionEntry.submissionSeed,
-      )
-      .consume(
-        lookupContract(transactionEntry, inputState),
-        lookupPackage(transactionEntry, inputState),
-        lookupKey(transactionEntry, inputState, knownKeys),
-      )
-      .fold(
-        err =>
-          reject(
-            recordTime,
-            buildRejectionLogEntry(transactionEntry, RejectionReason.Disputed(err.msg))),
-        _ => pass)
-  }
+        engine
+          .validate(
+            digest.abs,
+            digest.ledgerEffectiveTime,
+            ctx.getParticipantId,
+            digest.submissionTime,
+            digest.submissionSeed,
+          )
+          .consume(
+            lookupContract(digest, ctx.inputs),
+            lookupPackage(digest, ctx.inputs),
+            lookupKey(digest, ctx.inputs, knownKeys),
+          )
+          .fold(
+            err =>
+              reject[TransactionCommitter.Digest](
+                ctx.getRecordTime,
+                buildRejectionLogEntry(digest, RejectionReason.Disputed(err.msg))),
+            _ => StepContinue[TransactionCommitter.Digest](digest)
+          )
+      })
 
   /** Validate the submission's conformance to the DAML model */
-  private def authorizeAndBlind(
-      recordTime: Timestamp,
-      transactionEntry: TransactionEntry,
-  ): Commit[BlindingInfo] = delay {
-    Blinding
-      .checkAuthorizationAndBlind(
-        transactionEntry.abs,
-        initialAuthorizers = Set(transactionEntry.submitter),
-      )
-      .fold(
-        err =>
-          reject(
-            recordTime,
-            buildRejectionLogEntry(transactionEntry, RejectionReason.Disputed(err.msg))),
-        pure)
-  }
+  private def authorizeAndBlind: Step =
+    (ctx, digest) =>
+      Blinding
+        .checkAuthorizationAndBlind(
+          digest.abs,
+          initialAuthorizers = Set(digest.submitter),
+        )
+        .fold(
+          err =>
+            reject(
+              ctx.getRecordTime,
+              buildRejectionLogEntry(digest, RejectionReason.Disputed(err.msg))),
+          succ => buildFinalResult(ctx, digest, succ))
+
+  private def validateContractKeys: Step = (ctx, digest) => {
+    val damlState = ctx.inputs.collect { case (k, Some(v)) => k -> v } ++ ctx.getOutputs
+    val startingKeys = damlState.collect {
+      case (k, v) if k.hasContractKey && v.getContractKeyState.getContractId.nonEmpty => k
+    }.toSet
+    validateContractKeyUniqueness(ctx.getRecordTime, digest, startingKeys) match {
```

Comment:
Might be worth implementing `flatMap` on `StepResult`.

Reply:
I could also implement semigroup :-) I think it is not worth it for just one occurence